packages/apollo-gateway/src/buildQueryPlan.ts

```diff
@@ -545,17 +550,86 @@ function completeField(
 
     parentGroup.otherDependentGroups.push(...subGroup.dependentGroups);
 
+    let definition: FragmentDefinitionNode;
+    let selectionSet: SelectionSetNode;
+
+    if (context.compressDownstreamRequests && subGroup.fields.length > 2) {
+      ({ definition, selectionSet } = getInternalFragment(
+        subGroup.fields,
+        returnType,
+        context,
+      ));
+      parentGroup.internalFragments.add(definition);
+    } else {
+      selectionSet = selectionSetFromFieldSet(subGroup.fields, returnType);
+    }
+
+    // "Hoist" internalFragments of the subGroup into the parentGroup so all
+    // fragments can be included in the final request for the root FetchGroup
+    subGroup.internalFragments.forEach(fragment => {
+      parentGroup.internalFragments.add(fragment);
+    });
+
     return {
       scope,
       fieldNode: {
         ...fieldNode,
-        selectionSet: selectionSetFromFieldSet(subGroup.fields, returnType),
+        selectionSet,
       },
       fieldDef,
     };
   }
 }
 
+function getInternalFragment(
+  fields: Field<GraphQLCompositeType>[],
+  returnType: GraphQLCompositeType,
+  context: QueryPlanningContext
+) {
+  const expandedSelectionSet = selectionSetFromFieldSet(fields, returnType);
+
+  const key = print(expandedSelectionSet);
```

Comment:
Printers aren't known for performance, is there anything else we could use as a key here?  Say, the output of a `toConfig()`?

Reply:
`toConfig()` doesn't quite apply here since that's for schema nodes rather than document nodes, however I've gone with a `JSON.stringify` for now in 
03d3032.